### PR TITLE
Fix binary writing of local name indexes

### PR DESCRIPTION
**Pull request by kripken**
When writing a binary, we take the local indexes in the IR and turn
them into the format in the binary, which clumps them by type. When
writing the names section we should be aware of that ordering, but
we never were, as noticed in https://github.com/WebAssembly/binaryen/issues/3499

This fixes that by saving the mapping of locals when we are emitting
the name section, then using it when emitting the local names.

This also fixes the order of the types themselves as part of the
refactoring. We used to depend on the ordering of types to decide
which to emit first, but that isn't good for at least two reasons. First,
it hits https://github.com/WebAssembly/binaryen/issues/3648 - that order is not fully
defined for recursive types. Also, it's not good for code size - we've
ordered the locals in a way we think is best already (ReorderLocals pass).
This PR makes us pick an order of types based on that, as much as
possible, that is, when we see a type for the first time we append it to
a list whose order we use.

Test changes: Some are just because we use a different order than
before, as in `atomics64`. But some are actual fixes, e.g. in `heap-types`
where we now have ` (local $tv (ref null $vector))` which is indeed
right - `v` there is for `vector`, and likewise `m` for `matrix` etc. - we
just had wrong names before. Another example, we now have
`(local $local_externref externref)` whereas before the name was
`funcref`, and which was wrong... seems like the incorrectness was
more common on reference types and GC types, which is why this was
not noticed before.

Fixes https://github.com/WebAssembly/binaryen/issues/3499

Makes part of https://github.com/WebAssembly/binaryen/issues/3648 moot.

### CODE REVIEW

File: src/wasm/wasm-binary.cpp [hunk 1]
```diff
@@ -335,10 +335,18 @@ void WasmBinaryWriter::writeFunctions() {
     // Emit Stack IR if present, and if we can
     if (func->stackIR && !sourceMap && !DWARF) {
       BYN_TRACE("write Stack IR\n");
-      StackIRToBinaryWriter(*this, o, func).write();
+      StackIRToBinaryWriter writer(*this, o, func);
+      writer.write();
+      if (debugInfo) {
+        funcMappedLocals[func->name].swap(writer.getMappedLocals());
```
Comment (tlively) [inline src/wasm/wasm-binary.cpp:341]: Using `std::move` might be clearer here; `swap` makes it look like there might be something useful in the `funcMappedLocals` already. Same below.

File: src/wasm/wasm-binary.cpp [hunk 2]
```diff
@@ -663,20 +671,28 @@ void WasmBinaryWriter::writeNames() {
...
         assert(localsWithNames.size());
-        o << U32LEB(indexedFunc.first);
+        o << U32LEB(index);
         o << U32LEB(localsWithNames.size());
         for (auto& indexedLocal : localsWithNames) {
-          o << U32LEB(indexedLocal.first);
-          writeEscapedName(indexedLocal.second.str);
+          auto indexInFunc = indexedLocal.first;
+          auto name = indexedLocal.second;
+          // TODO: handle multivalue
```
Comment (tlively) [inline src/wasm/wasm-binary.cpp:691]: Do you plan to implement this soon or leave it for later?

**Review by tlively — COMMENTED** [review]
Would you mind undoing the change from `std::map` to `std::unordered_map` in this PR? I would like to see the changes in the tests that are due just to the other changes.

**Comment by tlively** [comment]
I find this code to be generally hard to understand (even though I wrote some of it myself) because of all the maps floating around as ambient state. Would it be a lot of extra work to split up the fixing of the names from the change to the local order? That might make it easier to review.

Also, I'm not sure new ordering is necessarily more principled than the old ordering. For example, there might be one extremely commonly accessed i32 local, but many more extremely rarely accessed i32 locals. It is possible that it might be better to split up the i32 locals into non-contiguous ranges. Assuming we don't want to split types into non-contiguous ranges of locals, though, it would be better to order types by the cumulative number of accesses to their locals rather than the maximum number of accesses to their locals.

File: src/wasm/wasm-binary.cpp (hunk 2 above)
Comment (kripken) [reply src/wasm/wasm-binary.cpp:691]: This is basically just writing out a TODO from earlier that wasn't spelled out - this never worked.

I don't have plans to work on it. I'm not sure offhand if anything can be done - the extra temp locals are not "normal", really, and may vanish in stacky form?

**Comment by kripken** [comment]
> Would you mind undoing the change from std::map to std::unordered_map in this PR? I would like to see the changes in the tests that are due just to the other changes.

That part of the change has no effects. We stop iterating on the data structure, so I changed it to unordered to basically reflect that. Keeping it ordered would be confusing, in other words, but have no noticeable effect.

> I find this code to be generally hard to understand (even though I wrote some of it myself) because of all the maps floating around as ambient state. Would it be a lot of extra work to split up the fixing of the names from the change to the local order? That might make it easier to review.

I don't think it can be split up. Changing the local order would break the names section. That is, right now things "work" because we have glitches that happen to be coordinated in most cases. But fixing one will make the other much more noticeable. I think the only reasonable thing is to fix the order, and to avoid breaking things further, to fix the names section at the same time.

> Also, I'm not sure new ordering is necessarily more principled than the old ordering. For example, there might be one extremely commonly accessed i32 local, but many more extremely rarely accessed i32 locals. It is possible that it might be better to split up the i32 locals into non-contiguous ranges. Assuming we don't want to split types into non-contiguous ranges of locals, though, it would be better to order types by the cumulative number of accesses to their locals rather than the maximum number of accesses to their locals.

Definitely true, yes. This will be closer to the optimal order picked by ReorderLocals (in that before we were just "random"), but the problems you mention are possible, and we can consider possible solutions.

**Review by tlively — APPROVED** [review]

**Pull request merged**